Source/DafnyDriver/LocalTestLoggerEvents.cs

```diff
@@ -0,0 +1,423 @@
+    /// </summary>
+    public override event EventHandler<TestRunMessageEventArgs> DiscoveryMessage;
+
+    /// <summary>
+    /// Raised when discovered tests are received
+    /// </summary>
+    public override event EventHandler<DiscoveredTestsEventArgs> DiscoveredTests;
+
+    /// <summary>
+    /// Raised when test discovery is complete
+    /// </summary>
+    public override event EventHandler<DiscoveryCompleteEventArgs> DiscoveryComplete;
+
+    #endregion
+
+    #region IDisposable
+
+    /// <summary>
+    /// Waits for all pending messages to be processed by the loggers cleans up.
+    /// </summary>
+    public void Dispose() {
+      if (this.isDisposed) {
+        return;
+      }
+
+      this.isDisposed = true;
+
+      // Ensure that the queue is processed before returning.
+      this.loggerEventQueue.Resume();
+      this.loggerEventQueue.Dispose();
+    }
+
+    #endregion
+
+    #region Internal Methods
+
+    /// <summary>
+    /// Enables sending of events to the loggers which are registered and flushes the queue.
+    /// </summary>
+    /// <remarks>
+    /// By default events are disabled and will not be raised until this method is called.
+    /// This is done because during logger initialization, errors could be sent and we do not
+    /// want them broadcast out to the loggers until all loggers have been enabled.  Without this
+    /// all loggers would not receive the errors which were sent prior to initialization finishing.
+    /// </remarks>
+    internal void EnableEvents() {
+      this.CheckDisposed();
+
+      this.loggerEventQueue.Resume();
+
+      // Allow currently queued events to flush from the queue.  This is done so that information
+      // logged during initialization completes processing before we begin other tasks.  This is
+      // important for instance when errors are logged during initialization and need to be output
+      // to the console before we begin outputting other information to the console.
+      this.loggerEventQueue.Flush();
+    }
+
+    /// <summary>
+    /// Raises a test run message event to the enabled loggers.
+    /// </summary>
+    /// <param name="args">Arguments to be raised.</param>
+    internal void RaiseTestRunMessage(TestRunMessageEventArgs args) {
+      if (args == null) {
+        throw new ArgumentNullException(nameof(args));
+      }
+
+      this.CheckDisposed();
+
+      // Sending 0 size as this event is not expected to contain any data.
+      this.SafeInvokeAsync(() => this.TestRunMessage, args, 0, "InternalTestLoggerEvents.SendTestRunMessage");
+    }
+
+    internal void WaitForEventCompletion() {
+      this.loggerEventQueue.Flush();
+    }
+
+    /// <summary>
+    /// Raises a test result event to the enabled loggers.
+    /// </summary>
+    /// <param name="args">Arguments to to be raised.</param>
+    internal void RaiseTestResult(TestResultEventArgs args) {
+      ValidateArg.NotNull(args, nameof(args));
```

Comment:
The Dafny code base uses `Contract.Requires(args != null);` for this. The disadvantage of `Contract.Requires` is that we no longer are getting run-time checks for it. If we decide to make a change toward using `ValidateArg` instead, then I'd really like to avoid the `nameof(args)` parameter, which I find to be a misguided convention that just clutters up the program text.

Reply:
Ditto.